### DISCUSSION TITLE
Extract text with a layer filter?

### DISCUSSION THREAD

**Discussion by BMaxV** [pymupdf/PyMuPDF#3116] [Looking for help] [answered]
Hello!

This is a fantastic library and will already be very useful for what I do, thanks for making it!

Here is my special snowflake case I need some help with:

I have an engineering drawing and that I want to extract some information from. This works very well for the shapes. The info I want is on a specific layer and I can just create a new page and recreate the shapes there, this is covered in the docs. I merely have to check `path["layer"]` and skip the layers I don't want.

I can't do this with text? 

The solution I have now is to get the text via the dictionary and look for the color value. But that feels suboptimal. If you have a suggestion on how to solve this better, that would amazing!

Thank you!

**Comment by JorjMcKie** [comment, marked as answer]
There are two ways to achieve what you need:
1. If you want to undergo some tedious details you can extract the text via the low-level method [get_texttrace](https://pymupdf.readthedocs.io/en/latest/functions.html#Page.get_texttrace). This output also has the "layer" key. But as you will see, some comfort compared to `get_text()` is missing there.
2. Use the document's feature directly to overall activate a certain layer. Then **_everything_** in the PDF will be switched to that view: page renderings, text extractions, vector graphics, etc. This is method [set_layer_ui_config](https://pymupdf.readthedocs.io/en/latest/document.html#Document.set_layer_ui_config). Get information about available such configurations via [layer_ui_configs](https://pymupdf.readthedocs.io/en/latest/document.html#Document.layer_ui_configs). These methods are equivalent to what you would achieve in a supporting PDF viewer when clicking on a layer.
